src/critics/local_goal_critic.cpp

```diff
@@ -0,0 +1,109 @@
+#include "mppic/critics/local_goal_critic.hpp"
+
+#include <xtensor/xmath.hpp>
+#include <xtensor/xsort.hpp>
+
+namespace mppi::critics
+{
+
+void LocalGoalCritic::initialize()
+{
+  auto getParam = parameters_handler_->getParamGetter(name_);
+
+  getParam(consider_angle_, "consider_angle_", true);
+  getParam(angle_offset_, "angle_offset", 6);
+  getParam(angle_cost_power_, "angle_cost_power", 1);
+  getParam(angle_cost_weight_, "angle_cost_weight", 3.0);
+
+  getParam(consider_distance_, "consider_distance", true);
+  getParam(distance_goal_count_, "disance_goal_count", 2);
+  getParam(distance_offset_, "distance_offset", 6);
+  getParam(distance_cost_power_, "distance_cost_power", 2);
+  getParam(distance_cost_weight_, "distance_cost_weight", 3.0);
+
+  getParam(stop_usage_path_reached_ratio_, "stop_usage_path_reached_ratio", 0.35);
+}
+
+void LocalGoalCritic::score(models::CriticFunctionData & data)
+{
+  if(!enabled_) {
+    return;
+  }
+
+  auto path_points = xt::view(
+    data.path, xt::all(), xt::range(0, 2));
+
+  auto find_furthest_point = [&]() {
+      auto last_points_ext =
+        xt::view(data.trajectories, xt::all(), -1, xt::newaxis(), xt::range(0, 2));
+      auto distances = xt::norm_l2(last_points_ext - path_points, {2});
+      size_t max_id_by_trajectories = 0;
+      double min_distance_by_path = std::numeric_limits<double>::max();
+
+      for (size_t i = 0; i < distances.shape(0); i++) {
+        size_t min_id_by_path = 0;
+        for (size_t j = 0; j < distances.shape(1); j++) {
+          if (min_distance_by_path < distances(i, j)) {
+            min_distance_by_path = distances(i, j);
+            min_id_by_path = j;
+          }
+        }
+        max_id_by_trajectories = std::max(max_id_by_trajectories, min_id_by_path);
+      }
+      return max_id_by_trajectories;
+    };
+
+  size_t furthest_reached_path_point_idx;
+  if (data.furthest_reached_path_point) {
+    furthest_reached_path_point_idx = *data.furthest_reached_path_point;
+  } else {
+    furthest_reached_path_point_idx = find_furthest_point();
+  }
+
+  auto lower_distance_offset = std::min(furthest_reached_path_point_idx + distance_offset_, path_points.shape(0) - 1);
+  auto upper_distance_offset = std::min(lower_distance_offset + distance_goal_count_, path_points.shape(0) - 1);
+
+  auto angle_offset = std::min(
+    furthest_reached_path_point_idx + angle_offset_, path_points.shape(
+      0) - 1);
+
+  size_t threshold_idx = static_cast<size_t>(
+    static_cast<double>(data.path.shape(0) - 1) * stop_usage_path_reached_ratio_);
+
+  if (consider_distance_ && lower_distance_offset < threshold_idx) {
+    auto path_interval = xt::view(path_points, xt::range(lower_distance_offset, upper_distance_offset), xt::all());
+    auto trajectories_points = xt::view(
+      data.trajectories, xt::all(), xt::all(),
+      xt::newaxis(), xt::range(0, 2));
+    auto distance_to_furthest = xt::norm_l2(trajectories_points - path_interval, {2});
```

Comment:
I'm not sure why this makes sense, why compare all `N` trajectory points to some subsection of path points right outside of the trajectory window? The first half or more of the trajectory points are obviously going to have a high distance, but they're all going to be non-zero. Taking the mean of this is weighting distance of 2 things that will never align and should always be relative constant, except with cross-track error. I think a good cost function should be near zero when its not in effect but the steady state cost of this is never 0 (eg goal critic is 0 at the goal; ref trajectory critic should be 0 when its exactly following the path)

Reply:
It seems like a mean-shifted approximation for cross-track error that introduces non-zero error at all time intervals

Reply:
Looking back at the last ref trajectory critic, it looks like its similar logic as here, but didn't pick up on that before. What's your core aim to measure with this portion of the critic? Maybe starting from that I can offer a concrete suggestion.

For instance, if really what you want to measure is cross track error, you could measure that instead or approximate it. Or if you want some distance local goal, why measure based on all path points and not just the last one?

Reply:
Its also comparing alot to the path itself, not just a local goal, so this feels like it should belong in a reference path critic rather than in a `local goal` critic, since its not just about a local goal on the path, there's a great deal of analysis of a segment of the path itself.

Reply:
> Its also comparing alot to the path itself, not just a local goal, so this feels like it should belong in a reference path critic rather than in a local goal critic, since its not just about a local goal on the path, there's a great deal of analysis of a segment of the path itself.

Maybe change name would be sufficient.

> I'm not sure why this makes sense, why compare all N trajectory points to some subsection of path points right

I forgot to change this. it must be mean distance from last trajectories points to path points range